### PR TITLE
New turn lanes conception

### PR DESCRIPTION
**Pull request by ivanPyrohivskyi**
Main issue https://github.com/osmandapp/OsmAnd/issues/15930
Other issues https://github.com/osmandapp/OsmAnd/issues/17238 
https://github.com/osmandapp/OsmAnd/issues/16752#issuecomment-1592569809

### CODE REVIEW

File: OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java [hunk 1]
```diff
@@ -2021,4 +1981,184 @@ private void avoidKeepForThroughMoving(List<RouteSegmentResult> result) {
...
+			if ((ln & 1) > 0) {
+				int[] oneActiveLane = {lanes[k]};
+				if (hasAllowedLanes(oldTurnType.getValue(), oneActiveLane, 0, 0)) {
+					tp = TurnType.getPrimaryTurn(lanes[k]);
+				}
+				cnt++;
+			}
+		}
+		TurnType t = TurnType.valueOf(tp, leftSide);
+		if (cnt == 3 && TurnType.isSlightTurn(t.getValue())) {
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java:2158]: here we need to add comment what 3 means

File: OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java [hunk 2]
```diff
@@ -2021,4 +1981,184 @@ private void avoidKeepForThroughMoving(List<RouteSegmentResult> result) {
...
+		return false;
+	}
+
+	private boolean hasSharpOrReverseTurnLane(String turnLanesPrevSegm) {
+		if (turnLanesPrevSegm == null) {
+			return false;
+		}
+		int[] uniqTurnTypes = getUniqTurnTypes(turnLanesPrevSegm);
+		for (int lane : uniqTurnTypes) {
+			if (lane == TurnType.TSHL || lane == TurnType.TSHR || lane == TurnType.TU || lane == TurnType.TRU) {
```
Comment (vshcherb) [inline OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java:2063]: Same method as below could be added TurnType.sharpOrReverse(..)

File: OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java (hunk 1 above)
Comment (ivanPyrohivskyi) [reply OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java:2158]: Old code, moved from 
https://github.com/osmandapp/OsmAnd/blob/7d3994d0844a28b94e85aa3de28ff91ce984f94f/OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java#LL1446C17-L1446C17

File: OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java [hunk 3]
```diff
@@ -2021,4 +1981,184 @@ private void avoidKeepForThroughMoving(List<RouteSegmentResult> result) {
...
+			add(TurnType.TSLL);
+		}};
+		Set<Integer> rightDirections = new HashSet<Integer>(){{
+			add(TurnType.TR);
+			add(TurnType.TSHR);
+			add(TurnType.TSLR);
+		}};
+		Set<Integer> sharpRightDirections = new HashSet<Integer>(){{
+			add(TurnType.TSHR);
+			add(TurnType.TRU);
```
[the remark this answers is not in the document]
Comment (ivanPyrohivskyi) [reply OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java:2111]: Here is compare with syntetic **mainTurnType** that calculated using angles. Just to know if we select right direction.

**Pull request merged**